README.md
# [LOOK FOR NEW MAINTAINER]

# build-timestamp-plugin
Adding `BUILD_TIMESTAMP` to jenkins env vars,

## Daylight Saving Time
DST is not enabled by default.
If your location uses DST, you can enable it by setting the `timezone`
to the correct `city` such as `America/New_York`.
Set the timezone to corresponding city, such as `America/New_York` to display Daylight Saving Time format when Daylight Saving Time is enabled.

## More vars and formats
Add more var names and formats if you need.

## Optional date/time shift
For additional variables you can define shift (days, hours, minutes) which will be added to build timestamp.
This allows to run build plan with previous day specified as parameter.

Export build timestamps to build env variables.

# Configure
`BUILD_TIMESTAMP` exported by default, and you can add more variables
with different format patterns in Global Configure page.  
![](docs/images/global-config.png)


# Using timestamps in Maven/Gradle/Shell
They are available in Maven build and Gradle build as built-in
properties/variables as well.

Maven:  
As defined property in pom, equals to

``` syntaxhighlighter-pre
System.getProperty('BUILD_TIMESTAMP')
```

Gradle:

``` syntaxhighlighter-pre
System.getenv('BUILD_TIMESTAMP')
```

Shell:

``` syntaxhighlighter-pre
"$BUILD_TIMESTAMP"
```

# Open Issues
[View these issues in Jira](https://issues.jenkins.io/issues/?jql=resolution%20is%20EMPTY%20and%20component%3D21120)



CHANGELOG.md
# Changes

## 1.0.3

-   [PR \#3](https://github.com/jenkinsci/build-timestamp-plugin/pull/3)

## 1.0.2

-   Added Optional date/time shift

    For additional variables you can define shift (days, hours, minutes)
    which will be added to build timestamp. This allows to run build
    plan with previous day specified as parameter.

## 1.0.1

-   Fixed validation issue (combobox does not support validation)

## 1.0.0

-   Initial release.

src/main/java/com/orctom/jenkins/plugin/buildtimestamp/Tuple.java
package com.orctom.jenkins.plugin.buildtimestamp;

import hudson.Extension;
import hudson.model.AbstractDescribableImpl;
import hudson.model.Descriptor;
import org.kohsuke.stapler.DataBoundConstructor;

/**
 * key value tuple
 * Created by hao on 12/16/15.
 */
public class Tuple extends AbstractDescribableImpl<Tuple> {

	private String key;
	private String value;

	@DataBoundConstructor
	public Tuple(String key, String value) {
		this.key = key;
		this.value = value;
	}

	public String getKey() {
		return key;
	}

	public String getValue() {
		return value;
	}

	@Extension
	public static class DescriptorImpl extends Descriptor<Tuple> {
		public String getDisplayName() { return ""; }
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		Tuple tuple = (Tuple) o;

		if (key != null ? !key.equals(tuple.key) : tuple.key != null) return false;
		return !(value != null ? !value.equals(tuple.value) : tuple.value != null);

	}

	@Override
	public int hashCode() {
		int result = key != null ? key.hashCode() : 0;
		result = 31 * result + (value != null ? value.hashCode() : 0);
		return result;
	}

	@Override
	public String toString() {
		return "[" + key + " -> " + value + "]";
	}
}


src/main/java/com/orctom/jenkins/plugin/buildtimestamp/BuildTimestampEnvironmentContributor.java
package com.orctom.jenkins.plugin.buildtimestamp;

import com.orctom.jenkins.plugin.buildtimestamp.BuildTimestampWrapper.DescriptorImpl;
import hudson.EnvVars;
import hudson.Extension;
import hudson.model.EnvironmentContributor;
import hudson.model.Run;
import hudson.model.TaskListener;
import jenkins.model.Jenkins;

import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.*;

import static com.orctom.jenkins.plugin.buildtimestamp.BuildTimestampPlugin.DEFAULT_PROPERTY;

/**
 * BuildTimestamp Environment Contributor
 * Created by hao on 12/16/15.
 */
@Extension
public class BuildTimestampEnvironmentContributor extends EnvironmentContributor {

	@Override
	public void buildEnvironmentFor(Run run, EnvVars envVars, TaskListener listener)
			throws IOException, InterruptedException {
		if (null == envVars.get(DEFAULT_PROPERTY)) {
			try {
				Map<String, String> timestampProperties = buildTimestamp(run.getTimestamp());
				envVars.putAll(timestampProperties);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}

	private Map<String, String> buildTimestamp(Calendar timestamp) {
		Map<String, String> timestampProperties = new HashMap<String, String>();
		DescriptorImpl descriptor = getDescriptorImpl();
		if (descriptor.isEnableBuildTimestamp()) {
			TimeZone timeZone = TimeZone.getTimeZone(descriptor.getTimezone());
			String pattern = descriptor.getPattern();

			setTimestamp(timestampProperties, DEFAULT_PROPERTY, format(timestamp, timeZone, pattern));

			Set<Tuple> extraProperties = descriptor.getExtraProperties();
			for (Tuple property : extraProperties) {
				setTimestamp(timestampProperties, property.getKey(), format(timestamp, timeZone, property.getValue()));
			}
		}
		return timestampProperties;
	}

	private void setTimestamp(Map<String, String> timestampProperties, String key, String value) {
		timestampProperties.put(key, value);
		System.setProperty(key, value);
	}

	private String format(Calendar timestamp, TimeZone timeZone, String pattern) {
		SimpleDateFormat format = new SimpleDateFormat(pattern);
		format.setTimeZone(timeZone);
		return format.format(timestamp.getTime());
	}

	public DescriptorImpl getDescriptorImpl() {
		return (DescriptorImpl) Jenkins.getInstance().getDescriptorOrDie(BuildTimestampWrapper.class);
	}
}


src/main/java/com/orctom/jenkins/plugin/buildtimestamp/BuildTimestampWrapper.java
package com.orctom.jenkins.plugin.buildtimestamp;

import hudson.Extension;
import hudson.model.AbstractProject;
import hudson.model.Descriptor;
import hudson.tasks.BuildWrapper;
import hudson.tasks.BuildWrapperDescriptor;
import hudson.util.ComboBoxModel;
import hudson.util.FormValidation;
import net.sf.json.JSONArray;
import net.sf.json.JSONObject;
import org.jenkinsci.Symbol;
import org.kohsuke.stapler.DataBoundSetter;
import org.kohsuke.stapler.QueryParameter;
import org.kohsuke.stapler.StaplerRequest;

import java.text.SimpleDateFormat;
import java.util.*;

import static com.orctom.jenkins.plugin.buildtimestamp.BuildTimestampPlugin.DEFAULT_PATTERN;

/**
 * wrapper
 * Created by hao on 12/16/15.
 */
public class BuildTimestampWrapper extends BuildWrapper {

	@Extension
	@Symbol("buildTimestamp")
	public static final class DescriptorImpl extends BuildWrapperDescriptor {

		private boolean enableBuildTimestamp = true;
		private String timezone = TimeZone.getDefault().getID();
		private String pattern = DEFAULT_PATTERN;
		private Set<BuildTimestampExtraProperty> extraProperties = new HashSet<BuildTimestampExtraProperty>();

		public DescriptorImpl() {
			load();
		}

		@Override
		public boolean isApplicable(AbstractProject<?, ?> abstractProject) {
			return false;
		}

		/**
		 * This human readable name is used in the configuration screen.
		 */
		public String getDisplayName() {
			return "";
		}

		@Override
		public boolean configure(StaplerRequest req, JSONObject formData) throws Descriptor.FormException {
			// reset optional authentication to default before data-binding
			this.enableBuildTimestamp = false;
			this.timezone = null;
			this.pattern = null;

			JSONObject data = formData.getJSONObject("enableBuildTimestamp");

			if (isNullJSONObject(data)) {
				enableBuildTimestamp = false;
			} else {
				enableBuildTimestamp = true;

				timezone = data.getString("timezone");
				pattern = data.getString("pattern");

				Object extraPropertyValues = data.get("extraProperties");
				if (null != extraPropertyValues) {
					extraProperties = extractExtraProperties(extraPropertyValues);
				} else {
					extraProperties = new HashSet<BuildTimestampExtraProperty>();
				}
			}

			save();
			return super.configure(req, formData);
		}

		private boolean isNullJSONObject(JSONObject data) {
			return null == data || data.isNullObject() || data.isEmpty();
		}

		private Set<BuildTimestampExtraProperty> extractExtraProperties(Object extraPropertyValues) {
			Set<BuildTimestampExtraProperty> properties = new HashSet<BuildTimestampExtraProperty>();
			if (extraPropertyValues instanceof JSONArray) {
				JSONArray array = (JSONArray) extraPropertyValues;
				for (Object item : array) {
					addProperty(properties, item);
				}
			} else if (extraPropertyValues instanceof JSONObject) {
				addProperty(properties, extraPropertyValues);
			}

			return properties;
		}

		private void addProperty(Set<BuildTimestampExtraProperty> properties, Object obj) {
			JSONObject data = (JSONObject) obj;
			String key = data.getString("key");
			String value = data.getString("value");
			String shiftExpression = data.getString("shiftExpression");
			if (isVariableNameValid(key) && isPatternValid(value) && ShiftExpressionHelper.isShiftExpressionValid(shiftExpression)) {
				properties.add(new BuildTimestampExtraProperty(key, value, shiftExpression));
			}
		}

		private boolean isVariableNameValid(String name) {
			return null != name && 0 != name.trim().length() && !name.matches(".*\\W.*");
		}

		private boolean isPatternValid(String pattern) {
			if (null == pattern || 0 == pattern.trim().length()) {
				return false;
			}
			try {
				new SimpleDateFormat(pattern);
				return true;
			} catch (Exception e) {
				return false;
			}
		}

		private String getConfiguredTimezone(String timezoneParam) {
			if (null == timezoneParam || 0 == timezoneParam.trim().length()) {
				return TimeZone.getDefault().getID();
			}

			return TimeZone.getTimeZone(timezoneParam).getID();
		}

		public FormValidation doCheckPattern(@QueryParameter("pattern") String patternParam,
											 @QueryParameter("timezone") String timezoneParam) {
			String configuredTimezone = getConfiguredTimezone(timezoneParam);
			String patternStr = DEFAULT_PATTERN;
			if (null != patternParam && 0 != patternParam.trim().length()) {
				patternStr = patternParam.trim();
			}

			try {
				SimpleDateFormat df = new SimpleDateFormat(patternStr);
				df.setTimeZone(TimeZone.getTimeZone(configuredTimezone));
				return FormValidation.ok("Using timezone: %s; Sample timestamp: %s", configuredTimezone, df.format(new Date()));
			} catch (IllegalArgumentException e) {
				return FormValidation.error("Invalid pattern");
			}
		}

		public FormValidation doCheckKey(@QueryParameter("key") String key) {
			if (isVariableNameValid(key)) {
				return FormValidation.ok();
			}
			return FormValidation.error("Invalid variable name");
		}

		public FormValidation doCheckValue(@QueryParameter("value") String value) {
			if (isPatternValid(value)) {
				return FormValidation.ok();
			}
			return FormValidation.error("Invalid pattern");
		}

		public FormValidation doCheckShiftExpression(@QueryParameter("shiftExpression") String shiftExpression) {
			if (ShiftExpressionHelper.isShiftExpressionValid(shiftExpression)) {
				return FormValidation.ok();
			}
			return FormValidation.error("Invalid time shift expression");
		}

		public ComboBoxModel doFillTimezoneItems() {
			ComboBoxModel items = new ComboBoxModel();
			Collections.addAll(items, TimeZone.getAvailableIDs());
			return items;
		}

		public boolean isEnableBuildTimestamp() {
			return enableBuildTimestamp;
		}

        @DataBoundSetter
        public void setEnableBuildTimestamp(boolean enableBuildTimestamp) {
            this.enableBuildTimestamp = enableBuildTimestamp;
        }

		public String getPattern() {
			return pattern;
		}

        @DataBoundSetter
        public void setPattern(String pattern) {
            this.pattern = pattern;
        }

		public String getTimezone() {
			return timezone;
		}

        @DataBoundSetter
        public void setTimezone(String timezone) {
            this.timezone = timezone;
        }

		public Set<BuildTimestampExtraProperty> getExtraProperties() {
			return extraProperties;
		}

        @DataBoundSetter
        public void setExtraProperties(Set<BuildTimestampExtraProperty> extraProperties) {
            this.extraProperties = extraProperties;
        }
	}
}
